apache/superset
# fix: Explore long URL problem

### SUMMARY
The objective of this PR is to fix the long URL problem In Explore. We had a function called `getExploreLongUrl` in `src/explore/exploreUtils` that was responsible for generating a URL with the contents of `form_data`. That was problematic because `form_data` can contain a lot of data which resulted in URL length-related errors. To fix this problem, https://github.com/apache/superset/pull/18151 created an endpoint called `/explore/form_data` where we can store this information on the server-side. This PR uses this endpoint to store `form_data` information and only keeps the generated key in the URL.

Another objective of the PR is to maintain backward compatibility with previously generated short URLs. We also want to generate new URLs every time the page loads to avoid content override by different users using the same shared URL. At the same time, after loading the page we want to preserve the generated URL and update its content as the user interacts with the page.

The `/explore` endpoint is using the old API and it contains a lot of complex logic related to the features in Explore so is NOT an objective of this PR to make big refactors that can potentially introduce regressions. I tried to preserve the original endpoint with the added feature. We should tackle this work in follow-up PRs, probably during the migration for the new API.

There's another function in `src/explore/exploreUtils` called `getExploreUrl` that transfers a pruned version of `form_data` using the URL and it's being used by the legacy charts and filter boxes. This will be tackled in another PR as it involves possible soon-to-be deprecated code.

We also have an endpoint called `/explore_json` that has the potential of being replaced by the new `/explore/form_data` endpoint but requires significant refactoring. This change is not in the scope of this PR.

### AFTER VIDEO

https://user-images.githubusercontent.com/70410625/151230992-e463f0ad-ce45-4420-894a-ff10ebb36508.mov

### TESTING INSTRUCTIONS
- Play with Explore controls and check if your state is preserved while refreshing or opening the URL in another tab
- Check if the actions that generate an URL are using the new `form_data_key` parameter
- Check if you can navigate from a dashboard to Explore without losing context

### ADDITIONAL INFORMATION
- [x] Has associated issue: closes #18198
